cpp/src_prims/sparse/distance.cuh

```diff
@@ -0,0 +1,366 @@
+
+namespace MLCommon {
+namespace Sparse {
+namespace Distance {
+
+template <typename value_idx, typename value_t>
+struct distances_config_t {
+  // left side
+  value_idx a_nrows;
+  value_idx a_ncols;
+  value_idx a_nnz;
+  value_idx *a_indptr;
+  value_idx *a_indices;
+  value_t *a_data;
+
+  // right side
+  value_idx b_nrows;
+  value_idx b_ncols;
+  value_idx b_nnz;
+  value_idx *b_indptr;
+  value_idx *b_indices;
+  value_t *b_data;
+
+  cusparseHandle_t handle;
+
+  std::shared_ptr<deviceAllocator> allocator;
+  cudaStream_t stream;
+};
+
+template <typename value_t>
+class distances_t {
+ public:
+  virtual void compute(value_t *out) { CUML_LOG_DEBUG("INside base"); }
+  virtual ~distances_t() = default;
+};
+
+/**
+ * Simple inner product distance with sparse matrix multiply
+ */
+template <typename value_idx = int, typename value_t = float>
+class ip_distances_t : public distances_t<value_t> {
+ public:
+  /**
+   * Computes simple sparse inner product distances as sum(x_y * y_k)
+   * @param[in] config specifies inputs, outputs, and sizes
+   */
+  explicit ip_distances_t(distances_config_t<value_idx, value_t> config)
+    : config_(config),
+      workspace(config.allocator, config.stream, 0),
+      csc_indptr(config.allocator, config.stream, 0),
+      csc_indices(config.allocator, config.stream, 0),
+      csc_data(config.allocator, config.stream, 0),
+      alpha(1.0) {
+    init_mat_descriptor(matA);
+    init_mat_descriptor(matB);
+    init_mat_descriptor(matC);
+    init_mat_descriptor(matD);
+
+    CUSPARSE_CHECK(cusparseCreateCsrgemm2Info(&info));
+
+    CUSPARSE_CHECK(
+      cusparseSetPointerMode(config.handle, CUSPARSE_POINTER_MODE_HOST));
+  }
+
+  /**
+   * Performs pairwise distance computation and computes output distances
+   * @param out_distances dense output matrix (size a_nrows * b_nrows)
+   */
+  void compute(value_t *out_distances) {
+    /**
+	   * Compute pairwise distances and return dense matrix in column-major format
+	   */
+
+    CUML_LOG_DEBUG("Compute() inside inner-product d");
+    device_buffer<value_idx> out_batch_indptr(config_.allocator, config_.stream,
+                                              config_.a_nrows + 1);
+    device_buffer<value_idx> out_batch_indices(config_.allocator,
+                                               config_.stream, 0);
+    device_buffer<value_t> out_batch_data(config_.allocator, config_.stream, 0);
+
+    value_idx out_batch_nnz = get_nnz(out_batch_indptr.data());
+
+    out_batch_indices.resize(out_batch_nnz, config_.stream);
+    out_batch_data.resize(out_batch_nnz, config_.stream);
+
+    compute_gemm(out_batch_indptr.data(), out_batch_indices.data(),
+                 out_batch_data.data());
+
+    /**
+     * Convert output to dense
+     * TODO: This is wasteful of memory and adds additional latency.
+     * It would be nice if there was a gemm that could do
+     * (sparse, sparse)->dense natively.
+     */
```

Comment:
I have't done extensive testing, but my nsight profiling has not revealed it to be one of the top bottlenecks. I have not profiled the dense conversion at massive scale, though, and I can imagine the parallelism would eventually create a bottleneck as the column sizes grow above the max allowable threads per block. Computing the GEMM on a CSR (which is often likely to have nnz = n*m) and converting to dense form also creates the need to allocate intermediate memory that we could (hopefully) be able to avoid if we were to just output directly to dense form.

Reply:
I asked this since dense conversion will always be quadratic in footprint and atleast quadratic in runtime.

Sounds like a nice little primitive for RAFT, someday in the future! Thanks Corey.

Reply:
> I asked this since dense conversion will always be quadratic in footprint and atleast quadratic in runtime.

This is true (though I believe it's actually going to be O(n*m) because the two sides of the computation can be different, full quadratic in the case where they are the same). The thing is, the sparse form is often going to be almost the same total number of elements, or only marginally smaller than, the dense form. This means we have to allocate the memory twice- once for the sparse output and then again for the dense output. To be clear- I think it would be very useful to have a gemm that would not require this extra memory. It would be even better if one side of the gemm didn't require the transpose, because that adds even more memory and compute overhead.

I've been designing a semiring primitive for the metrics that can make use of it. With sparse matrices this boils down to an intersection with the addition of the complement of the intersection of the columns for each pair of vectors. It can also be derived from a gemm with custom functions for the sum() and product(). All of the expanded lp metrics and their variants can be cast in this framework (and really, I believe any metric that is commutative and associative over the columns of only the pair.) 

That said, with all the additional memory use aside, my profiling did reveal that the cusparse gemm is the bottleneck, which doesn't surprise me at all. I think the semiring prim will save those additional memory allocations and compute expanded form metrics like L1, Canberra, and Linf much much faster than scipy/sklearn. I don't believe this will replace the gemm, though, since a dot product is really only concerned about the intersecting nonzero columns and a the generic semiring is concerned with both the intersection and its complement.

It would be nice to see a CSR gemm (maybe in raft?) that 1) would not require an explicit transpose and would output to dense form.

As for the semiring prim, I'm currently chunking the offsets and values for matrix B across a set of threads/blocks and having them perform the intersection with A. The chunking is balancing the load across the threads to keep them all as busy as possible. What would be really nice would be to have a semiring intersection primitive (e.g. gem with custom functions for sum() and product()) and an intersection complement primitive, that would accept the same functions and apply the sum() function to the output of the gemm. This way we could compute things like simple dot products in a single pass with the intersection primitive, but be able to reuse that to compute over the complement of the intersection. What do you think?

Reply:
I'd like to learn more about this use-case from you. Let's have a quick call sometime?

Reply:
Sounds good. Looks like something was scheduled for tomorrow.

Reply:
Also, here's a link to the original PR for the semiring prim. I'm not sure why it was closed (apparently I closed it), but the description should provide some useful background: https://github.com/rapidsai/cuml/pull/2861.